### PR TITLE
NetworkCloud 2023-05-01 API scenarios tests created for Networkcloud …

### PR DESCRIPTION
**Pull request by priyamshet**
### Related command
```
az networkcloud -h

Group
    az networkcloud : Manage Network Cloud resources.
        This command group is experimental and under development. Reference and support levels:
        https://aka.ms/CLI_refstatus
Subgroups:
    baremetalmachine     : Manage bare metal machine.
    cloudservicesnetwork : Manage cloud services network.
    cluster              : Manage cluster.
    clustermanager       : Manage cluster manager.
    defaultcninetwork    : Manage default CNI network.
    hybridakscluster     : Manage additional details of Hybrid AKS provisioned cluster.
    kubernetescluster    : Manage Kubernetes cluster.
    l2network            : Manage layer 2 (l2) network.
    l3network            : Manage layer 3 (l3) network.
    rack                 : Manage rack.
    racksku              : Manage rack SKU.
    storageappliance     : Manage storage appliance.
    trunkednetwork       : Manage trunked network.
    virtualmachine       : Manage virtual machine.
    volume               : Manage volume.

To search AI knowledge base for examples, use: az find "az networkcloud"
```
### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`. 

NetworkCLoud AAZ PR : https://github.com/Azure/aaz/pull/214

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

**Comment by priyamshet** [comment]
@kairu-ms   Can we get request your review on this PR

File: src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+        bmc_credentials = cls._args_schema.bmc_credentials
+        bmc_credentials.password = AAZStrArg(
+            options=["password"],
+            help="The password of the administrator of the device used during initialization.",
+            required=True,
+            fmt=AAZStrArgFormat(
+                min_length=1,
+            ),
+        )
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py:112]: We supported prompt input for password kinds of argument in aaz-dev-tools and azuire-cli. You can reference this [prompt-input-for-arguments doc](https://azure.github.io/aaz-dev-tools/pages/usage/workspace-editor/#prompt-input-for-arguments)

File: src/networkcloud/azext_networkcloud/operations/custom_properties.py [hunk 2]
```diff
@@ -112,14 +117,17 @@ def _build_schema_operation_status_result_read(cls, _schema):
             _schema.id = cls._schema_operation_status_result_read.id
             _schema.name = cls._schema_operation_status_result_read.name
             _schema.operations = cls._schema_operation_status_result_read.operations
-            _schema.percent_complete = cls._schema_operation_status_result_read.percent_complete
+            _schema.percent_complete = (
+                cls._schema_operation_status_result_read.percent_complete
+            )
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/custom_properties.py:122]:
```suggestion
            _schema.percent_complete = cls._schema_operation_status_result_read.percent_complete
```

File: src/networkcloud/azext_networkcloud/operations/virtualmachine/_create.py [hunk 3]
```diff
@@ -2,65 +2,47 @@
...
     def _build_arguments_schema(cls, *args, **kwargs):
         args_schema = super()._build_arguments_schema(*args, **kwargs)
 
         # Build VM Authentication args
-        args_schema.generate_ssh_keys = AAZBoolArg(
-            options=["--generate-ssh-keys"],
-            arg_group="Authentication",
-            help="Generate SSH public and private key files if missing. The keys will " +
-            "be stored under '~/.ssh'."
+        args_schema = CustomSshOptions.build_ssh_arg_schema(
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/virtualmachine/_create.py:34]: You can use `cls` as it inherited the `CustomSshOptions` class
```suggestion
        args_schema = cls.build_ssh_arg_schema(
```

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+
+class CustomSshOptions:
+    """Custom class providing common ssh operations"""
+
+    @staticmethod
+    def generate_ssh_keys():
+        """generates ssh keys in user .ssh folder"""
+
+        key_name = "id_rsa"
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/common_ssh.py:29]: I think the `id_rsa` name is commonly used. Will the generation overwrite the existing keys?

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        ssh_keys = []
+        if bool(args.generate_ssh_keys):
+            ssh_keys += CustomSshOptions.generate_ssh_keys()
+        if list(args.ssh_dest_key_path):
+            ssh_keys += CustomSshOptions.get_ssh_keys_from_path(
+                list(args.ssh_dest_key_path)
+            )
+        if list(args.ssh_key_values):
+            ssh_keys += CustomSshOptions.add_ssh_key_action(list(args.ssh_key_values))
+        return ssh_keys
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/common_ssh.py:144]:
```suggestion
    @classmethod
    def add_ssh_config(cls, args):
        ssh_keys = []
        if bool(args.generate_ssh_keys):
            ssh_keys += cls.generate_ssh_keys()
        if list(args.ssh_dest_key_path):
            ssh_keys += cls.get_ssh_keys_from_path(
                list(args.ssh_dest_key_path)
            )
        if list(args.ssh_key_values):
            ssh_keys += cls.add_ssh_key_action(list(args.ssh_key_values))
        return ssh_keys
```

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            + "provided or a directory. If a directory is provided will attempt to load all the "
+            + "public keys (.pub) found.",
+        )
+        args_schema.ssh_dest_key_path.Element = AAZStrArg()
+        return args_schema
+
+    @staticmethod
+    def add_ssh_config(args):
+        ssh_keys = []
+        if bool(args.generate_ssh_keys):
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/common_ssh.py:136]:
```suggestion
        if has_value(args.generate_ssh_keys) and args.generate_ssh_keys.to_serialized_data():
```

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            ssh_keys += CustomSshOptions.generate_ssh_keys()
+        if list(args.ssh_dest_key_path):
+            ssh_keys += CustomSshOptions.get_ssh_keys_from_path(
+                list(args.ssh_dest_key_path)
+            )
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/common_ssh.py:141]: You can use to_serialized_data() function, which will serialize an AAZValue to python build-in type value.
```suggestion
        if has_value(args.ssh_dest_key_path) and args.ssh_dest_key_path.to_serialized_data():
            ssh_keys += CustomSshOptions.get_ssh_keys_from_path(
                args.ssh_dest_key_path.to_serialized_data()
            )
```

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                    f"An invalid directory or key was provided: {path}"
+                )
+        return key_list
+
+    @staticmethod
+    def add_ssh_key_action(values):
+        """validate and add ssh key to the list"""
+
+        key_data = []
+        for key in values:
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/common_ssh.py:96]: If the `values` comes from to_serialized_data(), the key is string type by default.

File: src/networkcloud/azext_networkcloud/operations/kubernetescluster/_create.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        )
+        args.control_plane_node_configuration.ssh_public_keys = ssh_keys_control_plane
+
+        # initial agent pool configuration
+        for x in args.initial_agent_pool_configurations:
+            ssh_keys_initial_pool = []
+            if x._data.get("ssh_key_values") is not None:
+                ssh_keys_initial_pool += CustomSshOptions.add_ssh_key_action(
+                    (x._data["ssh_key_values"].values())
+                )
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/operations/kubernetescluster/_create.py:91]: Please don't use `_data` property. It's internal use only.
```suggestion
            if has_value(x.ssh_key_values) and x.ssh_key_values.to_serialized_data():
                ssh_keys_initial_pool += CustomSshOptions.add_ssh_key_action(
                    (x.ssh_key_values.to_serialized_data())
                )
```

Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/kubernetescluster/_create.py]: Updated

File: src/networkcloud/azext_networkcloud/operations/virtualmachine/_create.py (hunk 3 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/virtualmachine/_create.py]: Updated

File: src/networkcloud/azext_networkcloud/operations/custom_properties.py (hunk 2 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/custom_properties.py]: Updated.

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py (hunk 5 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/common_ssh.py]: updated.

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py (hunk 4 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/common_ssh.py:30]: Updated.

File: src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py (hunk 1 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py]: updated.

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py (hunk 6 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/common_ssh.py]: Updated

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py (hunk 8 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/common_ssh.py:97]: As the calling function could not be updated to use to_serialized_data() due to an error (details in previous comment), leaving this as this. please let us know if we missed anything.

**Comment by priyamshet** [comment]
@kairu-ms We have addressed most of your feedback, Can we request your re-review on this please.

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py (hunk 7 above)
Comment (kairu-ms) [reply src/networkcloud/azext_networkcloud/operations/common_ssh.py]: Can you share the code raise this error?
The `AAZList` has `to_serialized_data` function, but the python build-in `list` doesn't have `to_serialized_data` function. The output of `to_serialized_data` for `AAZList` is a python `list`.

File: src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py [hunk 10]
```diff
@@ -109,6 +109,9 @@ def _build_arguments_schema(cls, *args, **kwargs):
             fmt=AAZStrArgFormat(
                 min_length=1,
             ),
+            blank=AAZPromptInput(
```
Comment (kairu-ms) [inline src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py:112]: The `password` property is not `AAZPasswordArg` type. Maybe you should set its format as `password` in swagger like this https://github.com/Azure/azure-rest-api-specs/blob/6914c95db58c8dd037faf07bcc21475fc19c57f3/specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/CommonDefinitions.json#L2533-L2534

File: src/networkcloud/azext_networkcloud/operations/common_ssh.py (hunk 7 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/operations/common_ssh.py]: Error
This is what we see, please let us know if we are missing anything.

File: src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py (hunk 10 above)
Comment (priyamshet) [reply src/networkcloud/azext_networkcloud/aaz/latest/networkcloud/baremetalmachine/_replace.py:112]: Thanks Kai , we will address this change in the next preview API's version, our current version of API are approved

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**